### PR TITLE
OTLP Ingestion

### LINKED ISSUE #4981: OTLP native support
**Is your feature request related to a problem? Please describe.**

Currently, Cortex only supports Prometheus remote write protocol, with OTel is becoming more and more popular, more apps use OTel instrument SDK to export metrics but we can't push OTLP metrics to Cortex directly.

**Describe the solution you'd like**

Adding OTLPHttp exporter  native support.

**Describe alternatives you've considered**

Although we can use OTel Collector to transform OTLP to Prometheus remote wirte protocol, which means we need extra maintenance  an collector service.

### PR DESCRIPTION
**Pull request by friedrichg**
**What this PR does**: Implements OTLP ingestion. Based on the good work done by @Aneurysm9 🙇 

~This needs integration tests and unit tests~

**Which issue(s) this PR fixes**:
Fixes #4981

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/api/api.go [hunk 1]
```diff
@@ -260,6 +260,7 @@ func (a *API) RegisterRuntimeConfig(runtimeConfigHandler http.HandlerFunc) {
 func (a *API) RegisterDistributor(d *distributor.Distributor, pushConfig distributor.Config) {
 	distributorpb.RegisterDistributorServer(a.server.GRPC, d)
 
+	a.RegisterRoute("/api/v1/push/otlp", push.OTLPHandler(a.sourceIPs, a.cfg.wrapDistributorPush(d)), true, "POST")
```
Comment (yeya24) [inline pkg/api/api.go:263]: https://github.com/prometheus/prometheus/blob/main/web/api/v1/api.go#L388
Can we align with Prometheus OTLP metrics endpoint?

File: pkg/util/push/otlp.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/pmetric/pmetricotlp"
+
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/log"
+)
+
+// OTLPHandler is a http.Handler which accepts OTLP metrics.
+func OTLPHandler(sourceIPs *middleware.SourceIPExtractor, push Func) http.Handler {
```
Comment (yeya24) [inline pkg/util/push/otlp.go:23]: Can we try to utilize https://github.com/prometheus/prometheus/blob/main/storage/remote/write_handler.go#L204 as much as possible?

For example, can `DecodeOTLPWriteRequest` be used here?

**Review by yeya24 — COMMENTED** [review]
Great work. And thanks @Aneurysm9 for the initial implementation!

I am just reading https://prometheus.io/blog/2024/03/14/commitment-to-opentelemetry/ and wondering if we need to make changes accordingly to other features mentioned in the blog post. For example `Support UTF-8 metric and label names`.

File: pkg/api/api.go (hunk 1 above)
Comment (friedrichg) [reply pkg/api/api.go]: did /api/v1/otlp/v1/metrics like the docs say https://prometheus.io/docs/prometheus/latest/querying/api/#otlp-receiver. I hope that is ok

File: pkg/util/push/otlp.go (hunk 2 above)
Comment (friedrichg) [reply pkg/util/push/otlp.go:22]: done

**Comment by yeya24** [comment]
Nice. I think we still need unit tests and integration tests.

**Review by yeya24 — COMMENTED** [review]
Let's make sure we have changelog

**Comment by friedrichg** [comment]
PTAL @yeya24 @alanprot This is just the beginning. I think we can merge and do more tests/features in following PRs

**Review by yeya24 — APPROVED** [review]

**Pull request merged**